### PR TITLE
Add a possibility to have non filterable notifications (#3331)

### PR DESCRIPTION
**Pull request by vlo-rte**
- In release note :
  -  In chapter : Features
  -  Text : #3331 : Add a possibility to have non filterable notifications

### CODE REVIEW

File: src/test/api/karate/users/perimeters/getCurrentUserWithPerimeters.feature [hunk 1]
```diff
@@ -202,7 +206,7 @@ Feature: Get current user with perimeters (endpoint tested : GET /CurrentUserWit
     Then status 200
     And match response.userData.login == 'operator1_fr'
     And assert response.computedPerimeters.length == 2
-    And match response.computedPerimeters contains only [{"process":"process15","state":"state1","rights":"ReceiveAndWrite"}, {"process":"process15","state":"state2","rights":"ReceiveAndWrite"}]
+    And match response.computedPerimeters contains only [{"process":"process15","state":"state1","rights":"ReceiveAndWrite","filteringNotificationAllowed":true}, {"process":"process15","state":"state2","rights":"ReceiveAndWrite","filteringNotificationAllowed":true}]
```
Comment (freddidierRTE) [inline src/test/api/karate/users/perimeters/getCurrentUserWithPerimeters.feature:209]: add in this test a value false for filteringNotificationAllowed

File: services/users/src/main/java/org/opfab/users/services/UserServiceImp.java [hunk 2]
```diff
@@ -104,6 +134,63 @@ public boolean isEachStateUniqueInPerimeter(Perimeter perimeter){
...
+            for (String stateId : stateIds) {
+                if (!isFilteringNotificationAllowedForThisProcessState(currentUserWithPerimeters, processId, stateId))
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean isFilteringNotificationAllowedForThisProcessState(CurrentUserWithPerimeters currentUserWithPerimeters,
+                                                                     String processId, String stateId){
```
Comment (freddidierRTE) [inline services/users/src/main/java/org/opfab/users/services/UserServiceImp.java:183]: proposal for a more efficient code : 

-  Get first all the the filtering not allowed for a user via userPerimeter and put in an object 
-  Go through the settings to check if one is exiting in the object 

To be discuss

File: services/users/src/main/java/org/opfab/users/services/UserServiceImp.java [hunk 3]
```diff
@@ -48,6 +48,18 @@ public class UserServiceImp implements UserService {
     @Autowired
     private RabbitTemplate rabbitTemplate;
 
+    /** Retrieve user information from repository for the user **/
```
Comment (freddidierRTE) [inline services/users/src/main/java/org/opfab/users/services/UserServiceImp.java:51]: remove this comment and precise the method name if needed 
Do the same for existing method comments

File: services/users/src/test/java/org/opfab/users/model/CurrentUserWithPerimetersDataShould.java [hunk 4]
```diff
@@ -22,22 +22,28 @@ void testComputePerimeters(){
         Perimeter p1 = PerimeterData.builder().
                 id("perimeterKarate10_1_RR").
                 process("process10").
-                stateRights(new HashSet<>(Arrays.asList(new StateRightData("state1", RightsEnum.RECEIVE),
-                                                        new StateRightData("state2", RightsEnum.RECEIVE)))).
+                stateRights(new HashSet<>(Arrays.asList(new StateRightData("state1", RightsEnum.RECEIVE, true),
+                                                        new StateRightData("state2", RightsEnum.RECEIVE, true)))).
```
Comment (freddidierRTE) [inline services/users/src/test/java/org/opfab/users/model/CurrentUserWithPerimetersDataShould.java:26]: Add in this class a test to check merging of filter notification allowed (testMergeNotificationAllowed)

**Comment by freddidierRTE** [comment]
We should deal with the following use case : 
- Unselect a state in notification screen with operator1_fr
- Connect as admin an set this state as non filterable 
- Connect again with operator1_fr , try to change notification filter , it is not possible anymore

I suggest : set the checkbox  of the non filterable state to selected and save automatically the configuration in the back 

or may be better : set a specific visual to show to user the pb and encourage him to change the config 

To be discuss 


Add a small documentation in  https://opfab.github.io/documentation/current/reference_doc/#_notification_configuration

**Review by freddidierRTE — CHANGES_REQUESTED** [review]

File: services/users/src/main/java/org/opfab/users/services/UserServiceImp.java (hunk 3 above)
Comment (vlo-rte) [reply services/users/src/main/java/org/opfab/users/services/UserServiceImp.java:51]: I think the names of the functions are already explicit so I removed the comments.

File: src/test/api/karate/users/perimeters/getCurrentUserWithPerimeters.feature (hunk 1 above)
Comment (vlo-rte) [reply src/test/api/karate/users/perimeters/getCurrentUserWithPerimeters.feature:209]: Done.

File: services/users/src/test/java/org/opfab/users/model/CurrentUserWithPerimetersDataShould.java (hunk 4 above)
Comment (vlo-rte) [reply services/users/src/test/java/org/opfab/users/model/CurrentUserWithPerimetersDataShould.java:26]: Done.

File: services/users/src/main/java/org/opfab/users/services/UserServiceImp.java [hunk 5]
```diff
@@ -104,6 +134,63 @@ public boolean isEachStateUniqueInPerimeter(Perimeter perimeter){
...
+
+    private boolean isFilteringNotificationAllowedForThisProcessState(CurrentUserWithPerimeters currentUserWithPerimeters,
+                                                                     String processId, String stateId){
+
+        for (ComputedPerimeter computedPerimeter: currentUserWithPerimeters.getComputedPerimeters()) {
+
+            if ((computedPerimeter.getProcess().equals(processId)) && (computedPerimeter.getState().equals(stateId)) &&
+                (computedPerimeter.getFilteringNotificationAllowed() != null) && (!computedPerimeter.getFilteringNotificationAllowed()))
+                    return false;
+        }
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/users/src/main/java/org/opfab/users/services/UserServiceImp.java]: Done.

File: services/users/src/main/java/org/opfab/users/services/UserServiceImp.java (hunk 2 above)
Comment (vlo-rte) [reply services/users/src/main/java/org/opfab/users/services/UserServiceImp.java:176]: After discussion with Frederic, this remark will be done in another issue : https://github.com/opfab/operatorfabric-core/issues/3524

**Comment by vlo-rte** [comment]
> We should deal with the following use case :
> 
>     * Unselect a state in notification screen with operator1_fr
> 
>     * Connect as admin and set this state as non filterable
> 
>     * Connect again with operator1_fr , try to change notification filter , it is not possible anymore
> 
> 
> I suggest : set the checkbox of the non filterable state to selected and save automatically the configuration in the back
> 
> or may be better : set a specific visual to show to user the pb and encourage him to change the config
> 
> To be discuss
> 
> Add a small documentation in https://opfab.github.io/documentation/current/reference_doc/#_notification_configuration

After discussion, for this situation, when the user will navigate to the notification configuration screen, a popup will be displayed telling : 
"You are unsubscribed from one or more notifications for which unsubscribing is now prohibited.
 Notification(s) concerned: ...
 By clicking OK, you will be subscribed to these notifications again."

**Comment by freddidierRTE** [comment]
Add a line between "Notification concerned" and the list of notifications concerned
Set font-weight to "bold" for the list of notifications 

image

**Pull request merged**